File: python/y2020/day05.py
#%%
from typing import Tuple

import numpy as np

from utils import load

raw = load("day05.txt")

# %%
def binary(x: str, l: int, h: int, lowc: str, hic: str) -> int:
    for w in x:
        if w == lowc:
            h = int((l + h) / 2)
        elif w == hic:
            l = int((l + h + 1) / 2)
        else:
            raise ValueError("WTF")
    assert l == h
    return l


def process(x: str) -> Tuple[int, int]:
    return (binary(x[:7], 0, 127, "F", "B"), binary(x[-3:], 0, 7, "L", "R"))


def test1() -> int:
    out = 0
    for seat in raw:
        row, col = process(seat)
        if (id_ := row * 8 + col) > out:
            out = id_
    assert out == 935
    return out


# %%
def test2() -> None:
    arr = np.zeros(test1() + 1)
    for seat in raw:
        row, col = process(seat)
        arr[row * 8 + col] = 1

    assert np.where(arr == 0)[0][-1] == 743

File: rust/src/y2020/day05.rs
use std::collections::HashSet;
use std::iter::FromIterator;
/// This is basically a binary representation of a number.

type Parsed<'a> = &'a str;
pub fn parse<'a>(raw: &'a str) -> Vec<Parsed> {
    raw.split('\n').collect()
}

fn bin_to_dec(x: &str, zero: &char) -> usize {
    x.chars()
        .enumerate()
        .filter(|(_, c)| c == zero)
        .map(|(i, _)| 2usize.pow((x.len() - 1 - i) as u32))
        .sum()
}

fn process(x: &str) -> usize {
    let r = bin_to_dec(&x[..7], &'B');
    let c = bin_to_dec(&x[7..], &'R');
    8 * r + c
}

pub fn part1(raw: &Vec<Parsed>) -> usize {
    raw.iter().map(|seat| process(seat)).max().unwrap()
}

pub fn part2(raw: &Vec<Parsed>) -> usize {
    let seats: HashSet<usize> = HashSet::from_iter(raw.iter().map(|seat| process(seat)));
    let cand = seats
        .iter()
        .filter(|x| !seats.contains(&(*x + 1)) && seats.contains(&(*x + 2)))
        .map(|x| x + 1)
        .collect::<Vec<_>>();

    assert_eq!(cand.len(), 1);
    cand[0]
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::utils::*;
    #[test]
    fn test1() {
        assert_eq!(part1(&parse(&read(2020, "day05.txt"))), 935)
    }

    #[test]
    fn test2() {
        assert_eq!(part2(&parse(&read(2020, "day05.txt"))), 743);
    }
}
